src/isl.rs

```diff
@@ -0,0 +1,211 @@
+use crate::result::{
+    invalid_schema_error_raw, unresolvable_schema_error, IonSchemaError, IonSchemaResult,
+};
+use crate::system::{SharedTypeStore, TypeId};
+use crate::types::TypeDefinition;
+use ion_rs::value::owned::{OwnedElement, OwnedStruct};
+use ion_rs::value::{Element, Sequence, Struct, SymbolToken};
+use ion_rs::IonType;
+use std::convert::{TryFrom, TryInto};
+
+/// Represents a public facing API for schema constraints [IslConstraint] which stores IslTypeRef
+#[derive(Debug, Clone)]
+pub enum IslConstraint {
+    AllOf(Vec<IslTypeRef>),
+    Type(IslTypeRef),
+}
+
+/// Represents a public facing schema type [IslType] which can be converted to a solid TypeRef using TypeStore
+#[derive(Debug, Clone)]
+pub struct IslType {
+    name: Option<String>,
+    constraints: Vec<IslConstraint>,
+}
+
+impl IslType {
+    pub fn new(name: Option<String>, constraints: Vec<IslConstraint>) -> Self {
+        Self { name, constraints }
+    }
+
+    pub fn name(&self) -> &Option<String> {
+        &self.name
+    }
+
+    pub fn constraints(&self) -> &[IslConstraint] {
+        &self.constraints
+    }
+}
+
+/// Parse constraints inside an [OwnedStruct] to an [IslType]
+impl TryFrom<&OwnedStruct> for IslType {
+    type Error = IonSchemaError;
+
+    fn try_from(ion_struct: &OwnedStruct) -> Result<Self, Self::Error> {
+        let mut constraints = vec![];
+
+        // parses the name of the type specified by schema
+        let type_name: Option<String>=  match ion_struct.get("name") {
+            Some(name_element) => match name_element.as_str() {
+                Some(name) => Some(name.to_owned()),
+                None => { return Err(invalid_schema_error_raw("A type name is not string/symbol, if the value is any null, or the text of the symbol is not defined.")) }
```

Comment:
I'd simplify this to `type names must be a string or a symbol with defined text`. It would be even friendlier to break this into different error cases for `wrong type`, `null`, and `unknown text`, but we can worry about that later.